### PR TITLE
Make the `type_of` return a generic type for generators

### LINKED ISSUE #67651: Confusing/incorrect error message with incoherent implementations and async blocks
```Rust
use std::future::Future;

pub trait Bound {}

pub struct Error;

impl Bound for Error {}

pub struct Wrap;

impl<T: Bound> From<T> for Wrap {
    fn from(input: T) -> Self {
        Wrap
    }
}

impl<T: Into<Wrap>> From<T> for Error {
    fn from(input: T) -> Self {
        Error
    }
}

fn fail() -> impl Future<Output = Result<(), Error>> {
    async move {
        let a: Result<(), Error> = Ok(());
        let a: () = a?;
        Ok(())
    }
}
```

This fails with `type annotations required: cannot resolve Error: std::convert::From<Error>` (E0283) instead of the expected `conflicting implementations of trait std::convert::From<Error> for type Error` (E0119). Changing `fail` to 
```Rust
fn fail() -> Result<(), Error> {
    let a: Result<(), Error> = Ok(());
    let a: () = a?;
    Ok(())
}
``` 
i.e. eliminating the async block leads to the expected error message being produced.

[playground demonstration](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d5fe05a99c552d394a8a829f66cd308e)

### PR DESCRIPTION
**Pull request by Zoxc**
Fixes https://github.com/rust-lang/rust/issues/67651.

r? @nikomatsakis

### CODE REVIEW

**Review by nikomatsakis — APPROVED** [review]
Awesome :)

**Comment by nikomatsakis** [comment]
@bors r+

Thanks @Zoxc

**Comment by nikomatsakis** [comment]
Could you check, @Zoxc, if this fixes https://github.com/rust-lang/rust/issues/66312 too?

**Comment by Zoxc** [comment]
@nikomatsakis It fixes the `async` case, but not the other.

**Comment by nikomatsakis** [comment]
OK, I missed the second case, I suspect that should be broken into its own issue.

**Comment by Dylan-DPC** [comment]
Failed in rollup 

@bors r-

**Comment by Dylan-DPC** [comment]
Failed in rollup 

@bors r-

**Comment by Zoxc** [comment]
Looks like the region variables are different on `i686-gnu` for some reason. It doesn't happen on `i686-pc-windows-msvc`, so it's probably not a 32-bit thing. We probably shouldn't be showing these to the user anyway.

**Comment by nikomatsakis** [comment]
I wonder if, once https://github.com/rust-lang/rust/pull/69189 lands, this problem will go away, @Zoxc

**Comment by estebank** [comment]
#69189 has landed. Can we rebase and retry?

**Comment by Zoxc** [comment]
It doesn't seem like https://github.com/rust-lang/rust/pull/69189 helps.

**Comment by Zoxc** [comment]
It appears the output from 64-bit builds now matches 32-bit builds. Not sure if that's due to https://github.com/rust-lang/rust/pull/69189. I expected that to erase regions.

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/70198#issuecomment-601962257, @bors r-

**Comment by nikomatsakis** [comment]
Specifically I think that this PR requires a `x.py --bless --compare-mode=nll` run

**Comment by nikomatsakis** [comment]
([relevant log output](https://dev.azure.com/rust-lang/rust/_build/results?buildId=23921&view=logs&j=99ab81d4-3e3b-5299-a10e-bbb9b4eeeaf0&t=a69c7444-70d3-5df2-0b72-c05906927b5f&l=25968))

File: src/librustc_typeck/collect/type_of.rs [hunk 1]
```diff
@@ -188,12 +188,12 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
         Node::Field(field) => icx.to_ty(&field.ty),
 
         Node::Expr(&Expr { kind: ExprKind::Closure(.., gen), .. }) => {
-            if gen.is_some() {
-                return tcx.typeck_tables_of(def_id).node_type(hir_id);
-            }
```
Comment (eddyb) [inline src/librustc_typeck/collect/type_of.rs]: That's funny, I almost did the opposite last week, trying to make #69968 work.
(It was a bad idea and the solution was to fix the actual code that couldn't handle unsubstitued closure types)

**Pull request merged**